### DISCUSSION TITLE
an execution flow dyeing idea.

### DISCUSSION THREAD

**Discussion by Squirre17** [pwndbg/pwndbg#1848] [Ideas]
I recently had a good idea while debugging the kernel.

When there are multiple macro conditions and branches in the kernel, it can be frustrating to find the branch conditions and specific execution paths(especially in function which have multiple rets). One idea is to use `dye start` to indicate the beginning of dyeing and `dye end` to indicate the end of dyeing. Then, at the end of the function, we can use pwndbg to highlight the paths we have taken on the source code context (possibly using terminal background colors to highlight certain lines). This way, we can determine which branches we have traversed within the function, simplifying the analysis scope and improving vulnerability reproduction and analysis speed. (Alternatively, real-time rendering of the context could be implemented after executing `dye start`).

A possible implementation could be setting breakpoints at the assembly beginnings of all lines in the current function and determining if they are reached based on the breakpoints. gdb likely provides mapping between assembly and source code levels. For sure, this is a very rough idea.

**Comment by disconnect3d** [comment]
The implementation could also just do `nexti` repeatedly and do the dyeing this way.

I think the idea is very interesting and it would be awesome to have this kind of tracing implemented. I am not 100% sure if GDB is the best place for that, since e.g. other tools like IDA can do such dyeing for you.

However,, if someone would like to work on that or pay for it, I would be happy to accept the PR (or talk about an offer to develop that as a job).
